### PR TITLE
fix: Delay the cast from IAmazonDynamoDB to AmazonDynamoDBClient in the DocumentModel 

### PR DESCRIPTION
**Pull request by ashovlin**
## Description

**Before:** When initializing the document model `Table` object with an `IAmazonDynamoDB`, .NET/Standard/Core, we cast it right away to an `AmazonDynamoDBClient`. 
* Because we didn't have an synchronous HTTP client, we removed the public sync apis from the low-level service client.
* However we never removed the synchronous public `Table` APIs, which still rely on the internal sync low-level APIs internally.
* So to work around this, we kept the sync low-level APIs as `internal` on `AmazonDynamoDBClient`. This is why we need the cast, since the sync low-level APIs are no longer part of the public `IAmazonDynamoDB`.

**Now:** We delay the cast until right before we need to access the internal, low-level sync APIs.
* For users still calling the sync APIs in .NET/Standard/Core, you can avoid this by moving to the async equivalents where possible.
* If you're still relying on the sync APIs but aren't providing an actual `AmazonDynamoDBClient`, you'll now see a clearer `InvalidOperationException` instead of the `NullReferenceException` from the failed cast.

**Future:**
1. We can deprecate the `Table` sync-over-async APIs entirely. This would remove the need for this cast.
2. Since .NET 5 added back a sync HTTP client, we've discussed internally re-exposing sync low-level APIs where possible.

## Motivation and Context
DOTNET-6755 and #1589 (most of the way, until we do on of the _Future_ ideas from above)

This allows `Table` consumers to mock the `IAmazonDynamoDB` from .NET/Standard/Core as long as you stick to async public APIs.

## Testing
```
 dotnet test .\sdk\src\Services\DynamoDBv2\DynamoDBv2.sln
dotnet test .\sdk\AWSSDK.CoreAndCustomUnitTests.NetStandard.sln
```
PENDING - .NET Standard integ tests for DDB

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [X] My code follows the code style of this project
- [X] My change requires a change to the documentation
- [X] I have updated the documentation accordingly
- [X] I have read the **README** document
- [X] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License

- [X] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: sdk/src/Services/DynamoDBv2/Custom/DocumentModel/DocumentBatchGet.cs [hunk 1]
```diff
@@ -427,15 +427,22 @@ private Results GetAttributeItems()
...
-#else
         private static void CallUntilCompletion(IAmazonDynamoDB client, BatchGetItemRequest request, Results allResults)
-#endif
         {
             do
             {
-                var serviceResponse = client.BatchGetItem(request);
+#if AWS_ASYNC_API
+                // Cast the IAmazonDynamoDB to the concrete client instead, so we can access the internal sync-over-async methods
+                var internalClient = client as AmazonDynamoDBClient;
```
Comment (normj) [inline sdk/src/Services/DynamoDBv2/Custom/DocumentModel/DocumentBatchGet.cs:436]: nit: You can move the casting logic outside of the loop and only do it once.

File: sdk/src/Services/DynamoDBv2/Custom/DocumentModel/DocumentBatchWrite.cs [hunk 2]
```diff
@@ -469,15 +469,22 @@ private static int GetNumberOfWrites(BatchWriteItemRequest request)
...
-        private void CallUntilCompletion(BatchWriteItemRequest request, Dictionary<string, Dictionary<Key, Document>> documentMap, AmazonDynamoDBClient client)
-#else
         private void CallUntilCompletion(BatchWriteItemRequest request, Dictionary<string, Dictionary<Key, Document>> documentMap, IAmazonDynamoDB client)
-#endif
         {
             do
             {
-                var result = client.BatchWriteItem(request);
+#if AWS_ASYNC_API
+                // Cast the IAmazonDynamoDB to the concrete client instead, so we can access the internal sync-over-async methods
```
Comment (normj) [inline sdk/src/Services/DynamoDBv2/Custom/DocumentModel/DocumentBatchWrite.cs:477]: nit: Move the casting logic above the loop.

**Review by normj — APPROVED** [review]
A couple minor nits but change looks good.

File: sdk/src/Services/DynamoDBv2/Custom/DocumentModel/DocumentBatchGet.cs (hunk 1 above)
Comment (ashovlin) [reply sdk/src/Services/DynamoDBv2/Custom/DocumentModel/DocumentBatchGet.cs]: Fixed in https://github.com/aws/aws-sdk-net/pull/3388/commits/aa1e3273fae0ba8826c15a91902728713e06a1d4

File: sdk/src/Services/DynamoDBv2/Custom/DocumentModel/DocumentBatchWrite.cs (hunk 2 above)
Comment (ashovlin) [reply sdk/src/Services/DynamoDBv2/Custom/DocumentModel/DocumentBatchWrite.cs]: Fixed in https://github.com/aws/aws-sdk-net/pull/3388/commits/aa1e3273fae0ba8826c15a91902728713e06a1d4

**Comment by ashovlin** [comment]
Realized that I was using too wide of a condition `AWS_ASYNC_API`, which was casting unnecessarily for .NET Framework 4.7.2 even though it still has sync APIs we can use. Switched to `NETSTANDARD` in https://github.com/aws/aws-sdk-net/pull/3388/commits/aa1e3273fae0ba8826c15a91902728713e06a1d4

This also made some .NET Framework tests invalid, so I removed them.

**Review by normj — APPROVED** [review]

**Review by philasmar — APPROVED** [review]

**Review by normj — APPROVED** [review]

**Comment by ashovlin** [comment]
Dry-run e8acaff2-7604-4ade-a913-8ea4cc934cfe passed after adding https://github.com/aws/aws-sdk-net/pull/3388/commits/2f8a80ea5900768e8fe24d18d513fed4f61a085e

**Pull request merged**